azure/services/roleassignments/roleassignments.go

```diff
@@ -32,54 +34,65 @@ import (
 )
 
 const azureBuiltInContributorID = "b24988ac-6180-42a0-ab88-20f7382dd24c"
+const serviceName = "roleassignments"
 
 // RoleAssignmentScope defines the scope interface for a role assignment service.
 type RoleAssignmentScope interface {
 	azure.ClusterDescriber
-	RoleAssignmentSpecs() []azure.RoleAssignmentSpec
+	azure.AsyncStatusUpdater
+	RoleAssignmentSpecs() []azure.ResourceSpecGetter
 }
 
 // Service provides operations on Azure resources.
 type Service struct {
 	Scope RoleAssignmentScope
 	client
+	async.Reconciler
 	virtualMachinesClient        virtualmachines.Client
 	virtualMachineScaleSetClient scalesets.Client
 }
 
 // New creates a new service.
 func New(scope RoleAssignmentScope) *Service {
+	client := newClient(scope)
 	return &Service{
 		Scope:                        scope,
-		client:                       newClient(scope),
+		client:                       client,
 		virtualMachinesClient:        virtualmachines.NewClient(scope),
 		virtualMachineScaleSetClient: scalesets.NewClient(scope),
+		Reconciler:                   async.New(scope, client, client),
 	}
 }
 
 // Reconcile creates a role assignment.
 func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "roleassignments.Service.Reconcile")
 	defer done()
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
 	for _, roleSpec := range s.Scope.RoleAssignmentSpecs() {
-		switch roleSpec.ResourceType {
+		rs := roleSpec.(*RoleAssignmentSpec)
+		scope := fmt.Sprintf("/subscriptions/%s/", s.Scope.SubscriptionID())
+		rs.Scope = scope
+		contributorRoleDefinitionID := fmt.Sprintf("/subscriptions/%s/providers/Microsoft.Authorization/roleDefinitions/%s", s.Scope.SubscriptionID(), azureBuiltInContributorID)
+		rs.RoleDefinitionID = contributorRoleDefinitionID
+		switch rs.ResourceType {
 		case azure.VirtualMachine:
-			return s.reconcileVM(ctx, roleSpec)
+			return s.reconcileVM(ctx, rs)
 		case azure.VirtualMachineScaleSet:
-			return s.reconcileVMSS(ctx, roleSpec)
+			return s.reconcileVMSS(ctx, rs)
 		default:
-			return errors.Errorf("unexpected resource type %q. Expected one of [%s, %s]", roleSpec.ResourceType,
+			return errors.Errorf("unexpected resource type %q. Expected one of [%s, %s]", rs.ResourceType,
 				azure.VirtualMachine, azure.VirtualMachineScaleSet)
 		}
 	}
 	return nil
 }
 
-func (s *Service) reconcileVM(ctx context.Context, roleSpec azure.RoleAssignmentSpec) error {
+func (s *Service) reconcileVM(ctx context.Context, roleSpec *RoleAssignmentSpec) error {
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "roleassignments.Service.reconcileVM")
 	defer done()
-
 	spec := &virtualmachines.VMSpec{
 		Name:          roleSpec.MachineName,
 		ResourceGroup: s.Scope.ResourceGroup(),
```

Comment:
I think we can get this value through `roleSpec.ResourceGroupName()` to avoid a scope call, which will simplify the test cases.

Reply:
I had to make little change in the code of `reconcileVM` and `reconcileVMSS` -- > I renamed them as `getVMPrincipalID` and `getVMSSPrincipalID` as there were simply fetching principal id from the api. 

I am doing the following: 
1. Defined a `RoleAssignmentResourceType` method in the RoleAssignment scope which should return the resoource type ( vm or vmss ) 

2. If the resource type is `VM` -- call `getVMPrincipalID` that fetched the principal ID for VM case

3. If the resource type is `VMSS` -- call `getVMSSPrincipalID` that fetched the principal ID for VMSS case

Now, I am passing this `principalID` to the `RoleAssignmentSpecs` as an argument so that the value gets set at the  RoleAssignmentSpec in machine scope. 
Due to this -- I do not have access to get the ResourceGroupName .

Let me know what you think.